### PR TITLE
Stabilize patch-in-config (and prefer config over manifest)

### PR DESCRIPTION
**Pull request by jonhoo**
Tracking issue: https://github.com/rust-lang/cargo/issues/9269

---

This stabilizes the `patch-in-config` feature ([unstable entry](https://doc.rust-lang.org/cargo/reference/unstable.html#patch-in-config)) following the discussion in https://github.com/rust-lang/cargo/issues/9269#issuecomment-904913263.

As requested, this PR _also_ changes the precedence behavior such that a `[patch]` for the same dependency in both `.cargo/config.toml` and `Cargo.toml` prefers the patch from the configuration file over the one from the manifest, which matches the behavior of other overlapping configuration options. The corresponding test has also been updated to reflect this change in behavior.

### CODE REVIEW

File: src/doc/src/reference/unstable.md [hunk 1]
```diff
@@ -1227,31 +1226,6 @@ The supported values for `FREQUENCY` are 'always` and 'never', which control
 whether or not a message is printed out at the end of `cargo build` / `cargo check`.
 
 
-### patch-in-config
```
Comment (ehuss) [inline src/doc/src/reference/unstable.md:1230]: Can you add a section at the bottom of the file that indicates which version this is stabilized in (roughly follow the form of the other "stabilized" sections).

**Comment by ehuss** [comment]
Thanks @jonhoo!

@rfcbot fcp merge

File: src/doc/src/reference/unstable.md (hunk 1 above)
Comment (jonhoo) [reply src/doc/src/reference/unstable.md:1230]: Done in https://github.com/rust-lang/cargo/pull/9839/commits/e2777ee6aa0c92ac5893e39a01177053bd86b39f !

**Comment by ehuss** [comment]
🎉 
@bors r+

**Pull request merged**